controllers/topology/reconcile_state.go

```diff
@@ -62,17 +66,158 @@ func (r *ClusterReconciler) reconcileCluster(ctx context.Context, current, desir
+			return err
+		}
+	}
+
+	return nil
+}
+
+// createMachineDeployment creates a MachineDeployment and the corresponding Templates.
+func (r *ClusterReconciler) createMachineDeployment(ctx context.Context, md machineDeploymentTopologyState) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	if _, err := r.reconcileReferencedTemplate(ctx, reconcileReferencedTemplateInput{
+		desired: md.infrastructureMachineTemplate,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to create the %s %s", md.object.GroupVersionKind(), md.object.Name)
+	}
+
+	if _, err := r.reconcileReferencedTemplate(ctx, reconcileReferencedTemplateInput{
+		desired: md.bootstrapTemplate,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to create the %s %s", md.object.GroupVersionKind(), md.object.Name)
+	}
+
+	log.Info("creating", md.object.GroupVersionKind(), md.object.GetName())
+	if err := r.Client.Create(ctx, md.object.DeepCopy()); err != nil {
+		return errors.Wrapf(err, "failed to create the %s %s", md.object.GroupVersionKind(), md.object.Name)
+	}
+	return nil
+}
+
+// updateMachineDeployment updates a MachineDeployment. Also rotates the corresponding Templates if necessary.
+func (r *ClusterReconciler) updateMachineDeployment(ctx context.Context, clusterName string, currentMD, desiredMD machineDeploymentTopologyState) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	cleanupOldInfrastructureTemplate, err := r.reconcileReferencedTemplate(ctx, reconcileReferencedTemplateInput{
+		ref:     &desiredMD.object.Spec.Template.Spec.InfrastructureRef,
+		current: currentMD.infrastructureMachineTemplate,
+		desired: desiredMD.infrastructureMachineTemplate,
+		templateNamer: func() string {
+			// TODO: FIXME: Move the name generation into a shared func after #5070 has been merged
+			return names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-%s-infra-", clusterName, desiredMD.object.Name))
+		},
+		compatibilityChecker: checkReferencedObjectsAreCompatible,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to update the %s %s", currentMD.object.GroupVersionKind(), currentMD.object.Name)
+	}
+
+	cleanupOldBootstrapTemplate, err := r.reconcileReferencedTemplate(ctx, reconcileReferencedTemplateInput{
+		ref:     desiredMD.object.Spec.Template.Spec.Bootstrap.ConfigRef,
+		current: currentMD.bootstrapTemplate,
+		desired: desiredMD.bootstrapTemplate,
+		templateNamer: func() string {
+			// TODO: FIXME: Move the name generation into a shared func after #5070 has been merged
+			return names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-%s-bootstrap-", clusterName, desiredMD.object.Name))
+		},
+		compatibilityChecker: checkReferencedObjectsAreInTheSameNamespace,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to update the %s %s", currentMD.object.GroupVersionKind(), currentMD.object.Name)
+	}
+
+	// Check differences between current and desired MachineDeployment, and eventually patch the current object.
+	// TODO @fabrizio Do we also need the DeepCopy here?
```

Comment:
@fabrizio Do we also need the DeepCopy here?

Reply:
I have to this a little bit more about this and open a separated issue if required (in case we should modify the helper),

Reply:
I'm also not sure why we need it in the other cases to be honest :)